8.0 add stock picking package preparation line

OpenCode opened the pull request:
Add new module _stock_picking_package_preparation_line_ to manage details on stock package preparation. This module is useful for other modules.

OpenCode committed:
[ADD] Added basic classes and basic view for details to start to work on

OpenCode committed:
[IMP] On product onchange return a full description for the line

OpenCode committed:
[ADD] When a PackagePreparation is created, a PackagePreparationLine is created for every stock move in the picking added on the Package

OpenCode committed:
[ADD] When a PackagePreparation change, a PackagePreparationLine is created for every stock move in the picking if a new picking is added on the Package

OpenCode committed:
[IMP] Product quantity and unit of measure required if a product is selected on StockPickingPackagePreparationLine

OpenCode committed:
[IMP] When a product is selectd, unit of measure on StockPickingPackagePreparationLine is set

OpenCode committed:
[IMP] Adding a line with a product on PackagePreparationLine, a stock move and a stock picking are created, too

OpenCode committed:
[IMP] Set picking created automatically as "To Do" when it is created

OpenCode committed:
[ADD] Security access rule

OpenCode committed:
[FIX] When a stock move is deleted, the relative PackagePreparationLine is deleted, too

OpenCode committed:
[FIX] Picking type on new picking is get from existing picking on PackagePreparation if it is possible

OpenCode committed:
[IMP] Commented an if

OpenCode committed:
[IMP] Add a sequence handle on PackagePreparationLine

OpenCode committed:
[IMP] Stock Move created for StockPickingPackagePreparationLine withuot move on "Put in Pack" step on StockPickingPackagePreparation

OpenCode committed:
[ADD] Tests created for some cases of use

OpenCode committed:
[FIX] Control on auto-created picking state. Picking must be "assigned"

OpenCode committed:
[IMP] Move data (get from package preparation line) are returned from a function, now. So other module can inherit this flow to manipulate data for stock move creation

OpenCode committed:
[IMP] If quantity change on package, a backorder is used to set the right quantity on detail

OpenCode committed:
[IMP] Product an Product Uom are editable only if not exist a move line

eLBati committed:
FIX stock.picking.package.preparation.line write: quantity of ready pickings can be directly written
FIX view: keep picking lits as main data

eLBati committed:
remove mail thread dep

eLBati commented:
Thanks @OpenCode 
please see travis errors and https://github.com/OpenCode/stock-logistics-workflow/pull/2

eLBati committed:
FIX changing quantity for a 'ready' picking, the system will not generate a back order, but will just deliver the new quantity

eLBati committed:
REF _update_line_ids

eLBati committed:
FIX when making a partial transfer with picking wizard, update details quantity of pack preparation accordingly

OpenCode committed:
[FIX] Delete blank line to fix Travis Error:

======== Testing test_flake8 ========
./stock_picking_package_preparation_line/tests/test_package_preparation_line.py:89:1: W391 blank line at end of file

OpenCode commented:
@eLBati Pull merged and errors fixed. All green now.

laetitia-gangloff commented on stock_picking_package_preparation_line/__init__.py:
```diff
@@ -0,0 +1,22 @@
+#
+##############################################################################
+
+from . import model
```
can you rename model to models to follow odoo guidelines (https://www.odoo.com/documentation/8.0/reference/guidelines.html) ?

OpenCode replied:
Done :)

OpenCode committed:
[IMP] Path name based on OCA standard (models and views)

laetitia-gangloff commented on stock_picking_package_preparation_line/view/stock_picking_package_preparation_line.xml:
```diff
@@ -0,0 +1,41 @@
+                            <field name="name" />
+                            <field name="product_id"/>
+                            <field name="product_uom_qty" />
+                            <field name="product_uom" />
```
product_uom should be visible only for user in group : product.group_uom

laetitia-gangloff commented on stock_picking_package_preparation_line/view/stock_picking_package_preparation_line.xml:
```diff
@@ -0,0 +1,41 @@
+                                       attrs="{'readonly': [('move_id', '!=', False)]}"/>
+                                <field name="product_uom_qty"
+                                       attrs="{'required': [('product_id', '!=', False)]}"/>
+                                <field name="product_uom"
```
same remarks :
product_uom should be visible only for user in group : product.group_uom

laetitia-gangloff commented on stock_picking_package_preparation_line/views/stock_picking_package_preparation_line.xml:
```diff
@@ -0,0 +1,41 @@
+                            <field name="name" />
+                            <field name="product_id"/>
+                            <field name="product_uom_qty" />
+                            <field name="product_uom" />
```
product_uom should be visible only for user in group : product.group_uom

OpenCode replied:
Sure! Done!

laetitia-gangloff commented on stock_picking_package_preparation_line/views/stock_picking_package_preparation_line.xml:
```diff
@@ -0,0 +1,41 @@
+                                       attrs="{'readonly': [('move_id', '!=', False)]}"/>
+                                <field name="product_uom_qty"
+                                       attrs="{'required': [('product_id', '!=', False)]}"/>
+                                <field name="product_uom"
```
product_uom should be visible only for user in group : product.group_uom

OpenCode replied:
Done!

laetitia-gangloff commented on stock_picking_package_preparation_line/models/stock.py:
```diff
@@ -0,0 +1,39 @@
+                prep_line_model = self.env[
+                    'stock.picking.package.preparation.line'
+                    ]
+                prep_lines = prep_line_model.search([
```
Is it not better to do this in sudo ? (question!)
In my functional test, I installed purchase module. Purchase user can access to stock move and change quantity. But when I change quantity I get the following error :

AccessError

Sorry, you are not allowed to access this document. Only users with the following access level are currently allowed to do that:
- Warehouse/User
- Warehouse/Manager

(Document model: stock.picking.package.preparation.line)

OpenCode replied:
Mmh....I'm not sure this is a good idea! There are two security rules, yet. Why you need to do this as Administrator? We can think to add a particular security rule to extend purchase security to change package preparation line, too.

eLBati replied:
@OpenCode how do you suggest to fix the use case described by @laetitia-gangloff ?

OpenCode replied:
@eLBati I think the best thing to do is to add the access  to model stock.picking.package.preparation.line in an existing group. It's useless to create a group only for a model. But, what's the best group we can use to add this rule, too?

eLBati replied:
I suggest to add to `stock.picking.package.preparation.line` the same access rules as `stock.move`

laetitia-gangloff replied:
It is the same access than stock.move.
But in my test, I installed purchase module. And purchase module come with new acl on stock.move for purchase user.

eLBati replied:
On 16 September 2015 at 12:45, Laetitia Gangloff (ACSONE) <
notifications@github.com> wrote:

> It is the same access than stock.move.
> 
> But in my test, I installed purchase module. And purchase module come with
> new acl on stock.move for purchase user.
> 
> I see the problem now, thanks.
> So yes, I think operations on 'stock.picking.package.preparation.line' in
> this 'write' should be done with sudo()

OpenCode replied:
@laetitia-gangloff have you the same problem, yet?

OpenCode committed:
[FIX] Security group for product uom in package preparation details

laetitia-gangloff commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,190 @@
+             "quantity")
+    product_uom = fields.Many2one('product.uom')
+    sequence = fields.Integer()
+    note = fields.Text()
```
this fields seems unused ?

OpenCode replied:
The field note now is visible on form :)

laetitia-gangloff commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,190 @@
+             "will not generate a back order, but will just deliver the new "
+             "quantity")
+    product_uom = fields.Many2one('product.uom')
+    sequence = fields.Integer()
```
What is the usage of this sequence ?
I see no order on it.

OpenCode replied:
Sequence is used as handle widget on view to order manually the lines

laetitia-gangloff replied:
But without order is not usefull, the order is not magically kept ;)

OpenCode replied:
You're right. Sorry :) Order is based on sequence, now!

laetitia-gangloff commented on stock_picking_package_preparation_line/views/stock_picking_package_preparation_line.xml:
```diff
@@ -0,0 +1,41 @@
+                        </tree>
+                        <form>
+                            <group>
+                                <field name="product_id"
```
question : why it is not possible to change product or uom when move is set ?

laetitia-gangloff commented:
some little question :
Why it is not possible to change product or uom when move is set ?
Why the picking_type_id is not in the view ?

OpenCode committed:
[IMP] Note on view of stock picking package preparation line

OpenCode commented:
@laetitia-gangloff 

**Why it is not possible to change product or uom when move is set ?**

When a product is moved, I prefer to keep the some data on stock move. An user can only change the quantity. If you need to change product, you must delete a line and create a new one. Have you a better idea?

laetitia-gangloff commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,190 @@
+                    'picking_type_id': picking_type.id,
+                    }
+                picking = picking_model.create(picking_data)
+                for line, move_data in moves:
```
Is it not possible to set move lines in picking_data ?
'move_lines': [(0,0,{}), ...]

OpenCode replied:
No. I need to know the ID of new stock move to assign it to my package line.

OpenCode committed:
[FIX] Stock picking package preparation line order based on sequence field

laetitia-gangloff commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,190 @@
+                #       assigned
+                picking.action_confirm()
+                # ----- Show an error if a picking is not confirmed
+                if picking.state != 'confirmed':
```
My picking is in state assigned at this step (see in debugger). I think it can be ok, no ?

OpenCode replied:
When a picking is created by my module, it pass automatically this steps:

draft -> confirmed -> assigned

If in one of them there is a problem, an error appears to user

laetitia-gangloff replied:
What I try to say is :

At the moment of the test (picking.state != 'confirmed') the picking created by your module is already in state assigned.

laetitia-gangloff commented on stock_picking_package_preparation_line/views/stock_picking_package_preparation_line.xml:
```diff
@@ -0,0 +1,42 @@
+                                   nolabel="1" placeholder="Description"/>
+                            <field name="move_id" invisible="1"/>
+                            <field name="note" colspan="4"
+                                   nolabel="1" placeholder="Note"/>
```
hmmm... not sure it is the best place. In the following capture (with note) the difference between the description and the note is not really visible, what do you think ?

![image](https://cloud.githubusercontent.com/assets/7602419/9821413/72cdf250-58ba-11e5-9b7d-269f334f3b38.png)

OpenCode replied:
I think now is better

laetitia-gangloff commented:
@OpenCode When a product is moved, I prefer to keep the some data on stock move. An user can only change the quantity. If you need to change product, you must delete a line and create a new one. Have you a better idea?

I don't know. I asked the question, because on stock move I can change these informations. So I thought it was possible from here.

Another question, what is the usage of line without product_id and without move_id ? Product shouldn't be required ?

And the following use case seems strange...
- I put two picking in my package.
- I put two other picking in my package.
- I remove the two first picking.
  result : the details contains lines from the 4 pickings
  expected : only new lines and lines from pickings in my package
  Note : There is no problem when I put pack, because line are linked to move, but I think it is not right to keep them in the detail list.

I must stop for now, but I hope to continue on monday.
Thank you for this interesting module :)

OpenCode committed:
[FIX] Label for note and description on form of package preparation line

OpenCode commented:
@laetitia-gangloff 

> Another question, what is the usage of line without product_id and without move_id ? Product shouldn't be required ?

This lines are description line. In a module of l10n_italy we need to have description line in this class. Product isn't required. Only Description is required. If a line haven't product_id, the software doesn't create the stock move.

> I don't know. I asked the question, ...........

and

> And the following use case .........

For this cases, I'm studying what happen to quants if I delete/change a move line from my code.

> Thank you for this interesting module :)

Thank you, too, for your ideas and support

eLBati committed:
FIX Expected singleton: stock.picking.package.preparation.line()

eLBati committed:
IMP the line_ids update: now, whatever happens to package, _update_line_ids can be called and lines will be updated
Without this, for example, if you delete a picking from the package, the related line stays in package

OpenCode committed:
[FIX] Delete package preparation line from package preparation when the relative picking is not in the many2many relation yet

OpenCode committed:
[FIX] Context parameter skip package preparation line update to not loop change from stock move to preparation line

OpenCode committed:
[FIX] Typo in test

OpenCode committed:
[ADD] Test for change on stock move related with package preparation line

OpenCode committed:
[MRG] Merge from andrea4ever repository

OpenCode committed:
[ADD] Test for removal picking fro package preparation to check the right behavior of package preparation line

andrea4ever committed:
[ADD] Tests for package with a description line and a product line

OpenCode committed:
[FIX] Package Preparation line description is set as stock move name when the detail is created from picking

OpenCode committed:
[ADD] Test if Package Preparation line description is set as stock move name when the detail is created from picking

OpenCode committed:
[FIX] Create package preparation line from stock move only if move is not in cancel state

OpenCode committed:
[FIX] Flake8 error on travis CI for local variable not used

OpenCode committed:
[IMP] Extended test_move_and_picking_create_from_line with data check

eLBati commented:
Tests added. This is :+1: for me

andrea4ever commented:
LGTM :+1:

tafaRU commented on stock_picking_package_preparation_line/README.rst:
```diff
@@ -0,0 +1,39 @@
+=============
+
+ * No special setup
+
```
Please add **Usage** (with only **try on runbot**  button) and **Bug Tracker** sections.
See https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst as reference.

OpenCode replied:
Done!

tafaRU commented on stock_picking_package_preparation_line/__openerp__.py:
```diff
@@ -0,0 +1,42 @@
+
+{
+    'name': 'Stock Picking Package Preparation Line',
+    'version': '1.0',
```
Please update version according to https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#version-numbers

OpenCode replied:
Done

tafaRU commented on stock_picking_package_preparation_line/__openerp__.py:
```diff
@@ -0,0 +1,42 @@
+        'views/stock_picking_package_preparation_line.xml',
+        'security/ir.model.access.csv',
+    ],
+    'test': [],
```
Please remove this useless key.

OpenCode replied:
Done

tafaRU commented:
@OpenCode, thanks for this module.
Few nitpicking comments otherwise :+1:

OpenCode committed:
[FIX] Removed useless key in manifest

OpenCode committed:
[FIX] Version number

OpenCode committed:
[IMP] Usage and Bug Tracker sections on Readme

OpenCode commented:
@tafaRU All done and all green!

tafaRU commented on stock_picking_package_preparation_line/__openerp__.py:
```diff
@@ -0,0 +1,41 @@
+
+{
+    'name': 'Stock Picking Package Preparation Line',
+    'version': '8.0.0',
```
Version is still wrong, you have to set it as `8.0.1.0.0`
See https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#version-numbers as reference.

OpenCode replied:
Done!

tafaRU commented on stock_picking_package_preparation_line/README.rst:
```diff
@@ -0,0 +1,61 @@
+
+To use this module, you need to:
+
+* go to ...
```
Please remove three lines above as they are useless.

OpenCode replied:
"To use this module, you need to:" and "go to ..." ?

tafaRU replied:
Yes, thanks!

tafaRU commented on stock_picking_package_preparation_line/README.rst:
```diff
@@ -0,0 +1,61 @@
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback `here <https://github.com/OCA/
+stock-logistics-workflow/issues/new?body=module:%20
+stock_picking_package_preparation_line%0Aversion:%20
+8.0.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
Here you have to put the odoo version that is `8.0`.

OpenCode replied:
Done!

tafaRU commented on stock_picking_package_preparation_line/README.rst:
```diff
@@ -0,0 +1,61 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License
+
```
Equal signs are missing, please add them.
See https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst as reference.

OpenCode replied:
Done!

OpenCode committed:
[FIX] Module version in according with OCA standard

OpenCode committed:
[FIX] Right Odoo version in bug tracking link

OpenCode committed:
[FIX] Equal sign on module name
[FIX] Link target for license

OpenCode committed:
[FIX] Useless lines

tafaRU commented on stock_picking_package_preparation_line/__openerp__.py:
```diff
@@ -0,0 +1,41 @@
+    'name': 'Stock Picking Package Preparation Line',
+    'version': '8.0.1.0.0',
+    'author': 'Apulia Software srl,Odoo Community Association (OCA)',
+    'maintainer': 'Apulia Software srl',
```
Why? AFAIK the maintainer is OCA, as written at the bottom of README file.

OpenCode committed:
[FIX] OCA as maintainer

OpenCode commented:
@tafaRU All done and all green! Again! LOL!

gurneyalex commented on stock_picking_package_preparation_line/models/stock.py:
```diff
@@ -0,0 +1,56 @@
+            ('move_id', '=', self.id),
+            ])
+        for pack_line in pack_lines:
+            packs |= pack_line.package_preparation_id
```
suggestion:

```
packs = pack_lines.mapped('package.preparation_id')
```

and you can dispose of line 28

OpenCode replied:
Done. Thank you for suggestion.

gurneyalex commented on stock_picking_package_preparation_line/models/stock.py:
```diff
@@ -0,0 +1,56 @@
+        packs = self.env['stock.picking.package.preparation']
+        pack_line_model = self.env['stock.picking.package.preparation.line']
+        pack_lines = pack_line_model.search([
+            ('move_id', '=', self.id),
```
since the method seems to always be called within a `for move in self` loop, why not move this loop in get_packs?

You just have to change the search domain to ('move_id', 'in', self.ids)

OpenCode replied:
Ok. Done. Thank you

OpenCode replied:
I prefer to keep it separate because the function, if necessary, can also be called externally from other modules on a single move

gurneyalex commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,243 @@
+                    ).product_uom_qty = values['product_uom_qty']
+                self.move_id.with_context(
+                    skip_update_line_ids=True
+                    ).picking_id.action_assign()
```
can we have self.movel_id.with_context(....) stored once in a local variable?

OpenCode replied:
Sure. Done.

gurneyalex commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,243 @@
+                    line_model.create(line_vals)
+
+    @api.model
+    def create(self, values):
```
needs fixing: missing decorator  `@api.returns('self', lambda value: value.id)` (otherwise you're breaking the chain call from v7 api)

OpenCode replied:
Why I must use returns decorator? with API 8, Create returns a recordset (I know) but this is solved by decorator model, or not? I'm confused about this

gurneyalex replied:
if you call this code from a v7 method (using super()) you'll get a recorset, when the v7 method expects the id of the record.

OpenCode replied:
Ok. Done. Thank you

OpenCode committed:
[FIX] keep retro compatibility with v7 API

OpenCode committed:
[IMP] improvement for multiple similar calls

OpenCode committed:
[IMP] improvement for mapped field

OpenCode committed:
[FIX] Test: KeyError: 'preparation_id'

OpenCode committed:
[FIX] Sequence with default for package preparation line

eLBati commented:
Hello @gurneyalex is this mergeable?
